ir/src/lib.rs

```diff
@@ -168,6 +182,58 @@ impl AirIR {
     }
 }
 
+// === HELPERS ====================================================================================
+
+/// Returns an error if a public inputs section is not defined.
+fn validate_public_inputs(public_inputs: &PublicInputs) -> Result<(), SemanticError> {
+    if public_inputs.is_empty() {
+        return Err(SemanticError::MissingSection(
+            "Public Inputs section is missing".to_string(),
+        ));
+    }
+    Ok(())
+}
+
+/// Returns an error if a boundary constraints section is not defined.
+fn validate_boundary_constraints(
+    boundary_constraints: &BoundaryConstraints,
+) -> Result<(), SemanticError> {
+    if boundary_constraints.main_first().is_empty()
+        && boundary_constraints.main_last().is_empty()
+        && boundary_constraints.aux_first().is_empty()
+        && boundary_constraints.aux_last().is_empty()
+    {
+        return Err(SemanticError::MissingSection(
+            "Boundary Constraints section is missing".to_string(),
+        ));
+    }
+    Ok(())
+}
+
+/// Returns an error if a transition constraints section is not defined.
+fn validate_transition_constraints(
+    transition_constraints: &TransitionConstraints,
+) -> Result<(), SemanticError> {
+    if transition_constraints.main_constraints().is_empty()
+        && transition_constraints.aux_constraints().is_empty()
+    {
+        return Err(SemanticError::MissingSection(
+            "Transition Constraints section is missing".to_string(),
+        ));
+    }
+    Ok(())
+}
+
+/// Returns true if the source contains trace columns section.
+fn contains_trace_cols(source: &Vec<ast::SourceSection>) -> bool {
+    for section in source {
+        if matches!(section, ast::SourceSection::TraceCols(_)) {
+            return true;
+        }
+    }
+    false
+}
```

Comment:
Since we can get the information about whether we have trace columns or not for free (when we process them in the match statement inside one of the iterators in `from_source`), I don't think it's optimal to iterate over all the sections here. 

Instead of the current approach, I think it might be simpler to just declare a verification struct with some booleans for each of the sections that are required. They would be initialized to false, then we'd just update them when the section is processed during the declaration and constraint handling. Finally, after all the source sections are processed, we can report an error based on the values of these booleans.

For example, in `from_source` we could do something like this:

```
let validator = SourceValidator::new()

// process the declarations of identifiers first, using a single symbol table to enforce
// uniqueness.
for section in source {
            match section {
                ...
                ast::SourceSection::TraceCols(columns) => {
                    validator.exists("trace_columns");
                    ...
                }
                ...
           }
}

// then process the constraints & validate them against the symbol table.
...

validator.check();

...
```

We would call "validator.exists" for all required source sections. In the future, if we needed to handle other types of error processing, we could incorporate that into this easily.

Reply:
Fixed, thanks.